Add support for B.A.T.M.A.N. Advanced

T-X opened the pull request:
This adds support for the layer 2 mesh routing protocol
B.A.T.M.A.N. Advanced. "batadv" can be used to filter on batman-adv
packets. It also allows later filters to look at frames inside the
tunnel when both "version" and "type" are specified.

Documentation for the batman-adv protocol can be found at the following
locations:

https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/networking/batman-adv.rst
https://www.open-mesh.org/

Signed-off-by: Linus Lüssing \<linus.luessing@c0d3.blue\>

T-X commented:
One remark:  pcap/batadv_packet.h and pcap/batadv_legacy_packet.h are basically copies from the Linux kernel and are therefore GPLv2 licensed. Let me know if that's okay for you or if I should ask the other maintainers and contributors for a BSD-3-Clause (compatible) dual-license for these two header files.

infrastation commented:
Thank you for preparing this feature. Neither macOS nor FreeBSD like the new Linux-specific headers. As you mentioned, these are GPL, also the amount of filter-specific code seems to be much smaller than the amount of declarations. So it might be worth to consider if the new code could have just the declarations it needs, then it would be simpler to make these portable and have the same licence as the rest of libpcap. Other opinions are welcome.

guyharris commented:
> So it might be worth to consider if the new code could have just the declarations it needs, then it would be simpler to make these portable and have the same licence as the rest of libpcap. Other opinions are welcome.

My opinion can be summarized as "+1". :-)

I.e., no GPLed code, please.

guyharris commented on pcap/batadv_packet.h:
```diff
@@ -0,0 +1,631 @@
+#ifndef _UAPI_LINUX_BATADV_PACKET_H_
+#define _UAPI_LINUX_BATADV_PACKET_H_
+
+#include <asm/byteorder.h>
```
None of those three headers are available on all the operating systems on which libpcap operates.  (Note that one of those operating systems is Windows.)

Find a way not to use them.

guyharris commented on pcap/batadv_packet.h:
```diff
@@ -0,0 +1,631 @@
+/* SPDX-License-Identifier: (GPL-2.0 WITH Linux-syscall-note) */
```
No GPLed code, please.

guyharris commented:
Here's a tarball containing a patch, pcap/batadv_legacy_packet.h, and pcap/batadv_packet.h.  The patch is a variant of your change that compiles on macOS Catalina and Xcode 11.7, so it rips out enough Linuxisms to get it to compile there, at least.

It also gets rid of all #pragma pack stuff (not all compilers necessarily support it), instead declaring 2-byte and 4-byte integral values as arrays of `uint8_t` of the appropriate length.

Do those headers need to be part of the libpcap distribution, rather than being internal to the libpcap source?

[tarfile.gz](https://github.com/the-tcpdump-group/libpcap/files/5580882/tarfile.gz)

T-X commented:
Thanks for the great feedback!

Changelog v2:
* changes to batadv_{legacy_,}packet.h:
  * integrated @guyharris'  suggested changes: removal of #pragma packed(2), bitfield endianess #ifdefs, Linux specific includes, unused definitions, conversion to uint8_t only
  * further removed kernel doc comments and instead added a small note to consult the Linux kernel version for a complete definition
  * changed the license to BSD-3 for these new, tiny versions
  * moved these two header files from the pcap directory to the root directory
  * shortened the include guards, the UAPI directory for instance was Linux specific

T-X commented:
Changelog v3:
* removed unused variables "s" and "b1" in gen_batadv() to satisfy the AppVeyor CI / Visual Studio builds

infrastation commented:
Passed FreeBSD build this time, that's better. One other thing that would complement this feature is some tcpdump test(s) to flex the new keywords. Do you have enough batman packets with the right protocol fields?

T-X commented:
Changelog v4:
* changed offset parameter of gen_batadv_push_offset() from size_t to u_int and added according type cast on callers, to make AppVeyor CI / Visual Studio happy

T-X commented:
@infrastation sure, generating some test packets is no issue, can do that next weekend. According tests need to be added to the tcpdump repository in a separate commit over there, right?

infrastation commented:
Yes, that's correct, and the tests will not pass until this feature has been merged, but eventually it would be consistent.

T-X commented:
Some simple tests for tcpdump for the batman-adv version 15 packet types and their decoding offsets can be found here: https://github.com/the-tcpdump-group/tcpdump/pull/891

infrastation commented:
Thank you. For what it's worth, I have just looked through the proposed changes one more time and could not find any immediate issues to fix (which does not automatically mean everything is good, but still). Other reviewers are welcome.

T-X commented:
I'm a bit hesitant to ask, as I can see with all the frequent commits that everyone is very eagerly working on libpcap/tcpdump. But any chance to get this merged? Would help me adding it to the collaborative projects I would want to use this in. Which are usually reluctant in adding changes that are not upstream yet.

mcr commented:
Hi, @fenner , you've been down the pcap compiler space more than I, do you have any final comments before I merge this?

guyharris commented on grammar.y.in:
```diff
@@ -592,11 +593,41 @@ other:	  pqual TK_BROADCAST	{ CHECK_PTR_VAL(($$ = gen_broadcast(cstate, $1))); }
+				type = pcap_nametobatadvtype_v15($2);
+				break;
+			default:
+				YYABORT;
```
This should probably call `bpf_set_error()` to report that $1 is an unsupported batman-adv version.

T-X replied:
Thanks! I've added an according error message with bpf_set_error().

T-X commented:
Changelog v5:
* adding "static" attribute to functions "gen_batadv_check_version" and "gen_batadv_check_type" to fix "no previous prototype for function" warnings
* removing unnecessary CHECK_INT_VAL($1) and CHECK_PTR_VAL($2) in grammar.y.in in pbatadv, fixing an unsigned vs. signed check warning with CHECK_INT_VAL($1)

T-X commented:
Changelog v6:
* rebasing to current master (tiny conflict since 5f634cf36c in pcap-filter.manmisc.in, resolved it)

AiyionPrime commented:
@guyharris is this ready to merge? Or am I overlooking some remark?

infrastation commented:
If it helps, I have not learned the parser code sufficiently well yet, so far I can make sense of roughly half of the proposed C code changes.

AiyionPrime commented:
> Hi, @fenner , you've been down the pcap compiler space more than I, do you have any final comments before I merge this?

@mcr did you get reply on that?
I'd really like to see this in libpcap;
and I'm confident @T-X would be available on the matter, if there were some urgent changes to this PR before it could get merged.

mcr commented:
> @mcr did you get reply on that? I'd really like to see this in libpcap; and I'm confident @T-X would be available on the matter, if there were some urgent changes to this PR before it could get merged.

I did not, but if @guyharris is okay with it, then I am.

AiyionPrime commented:
@guyharris its been three months :/
Are you by any chance okay with these changes?

AiyionPrime commented:
@guyharris its been four months, could you give this another round of a review?

fxlb commented:
To have the new header files included in a release archive, they must be added in `Makefile.in` (`PUBHDR` or `HDR`).
(See the result of new target `make releasecheck` here:
"fatal error: batadv_packet.h: No such file or directory"
https://github.com/the-tcpdump-group/libpcap/pull/980/checks?check_run_id=9728030848)
@guyharris @mcr:  Is it `PUBHDR` or `HDR` in this case?

T-X commented:
Changelog v7:
* added batadv_legacy_packet.h and batadv_packet.h to Makefile.in's HDR variable, to fix "make releasecheck" and "make releasetar"

@fxlb I've added it to HDR and not PUBHDR. I don't think it's necessary to expose these header files to a user's system include directory. It's more similar to ieee80211.h, ppp.h or pcap-usb-linux-common.h for instance.

T-X commented:
Is there anything else I can do to progress this or to motivate for its adoption?

batman-adv is a quite popular mesh routing protocol in various non-profit wireless community networks. Like [Freifunk](https://freifunk.net/en/) (via the [Gluon](https://github.com/freifunk-gluon/) firmware) or [AlterMundi](https://altermundi.net/) (via the [LibreMesh](https://github.com/libremesh/) firmware). But also commercial vendors use it, like Datto or Plasma Cloud.

On embedded devices Wireshark/tshark is not an option, even though libwireshark has a batman-adv dissector, due to the limited amount of flash and RAM in that case. So having this more tunable and better performing capture method would help a lot, as especially in these open, non-profit wireless community networks various layer 2 traffic can happen on top.

Next, I want to use this in Gluon with [bpfcountd](https://github.com/lemoer/bpfcountd/) to create more detailed traffic statistics, to be able to better react to issues and to improve batman-adv through real-world data: https://github.com/freifunk-gluon/gluon/pull/2367. And by that improving the stability and the features users in these communities can use, as well as the sizes they can scale their networks to.

fxlb commented:
@guyharris: Do you think your comments have been taken into account? Do you think this PR is ready to be merged?

AiyionPrime commented:
@fxlb Have you gotten any feedback from @guyharris on this?
It's been another four months; is there anything I can provide to get this thing moving forward?

fxlb commented:
The current types/sub-types in libpcap filter syntax use "-" and not "_".

Examples:
       wlan:
       assoc-req,  assoc-resp, ...

       The  following   ICMP   type   field   values   are   available:
       icmp-echoreply,  icmp-unreach, ...

       The  following  ICMPv6  type   field   values   are   available:
       icmp6-destinationunreach, icmp6-packettoobig, ...

       The following TCP flags field  values  are  available:  tcp-fin,
       tcp-syn, tcp-rst, tcp-push, tcp-ack, tcp-urg, tcp-ece, tcp-cwr.

Thus for batman-adv packet types,
       The following packet type aliases are available for compat  ver-
       sion  14:  iv_ogm,  unicast-frag (?), tt_query, roam_adv, unicast_4addr.

       The following packet type aliases are available for compat  ver-
       sion 15: iv_ogm, unicast_frag, unicast_4addr, unicast_tvlv.

it should be better to replace "_" by "-".

T-X commented:
Changelog v6:
* rebased to current master (no changes)
* changed all packet sub-type aliases to use hyphens instead of underscores and updated manpage accordingly

T-X commented:
Changelog v7:
* rebased to current master branch
* added new batman-adv multicast packet type, introduced in batman-adv v2024.0

T-X commented:
I would also like to add the correct offset to the batadv_mcast_packet type, which has a variable payload offset. I would need to add the contents of the two bytes tvlv_len field to the offset:
https://git.open-mesh.org/batman-adv.git/blob/9e1f3974cca1fc0866793076684ea0f5c2649412:/include/uapi/linux/batadv_packet.h#l509

My tries so far were not successful with this, I might need some help/guidance for that.

T-X commented:
Changelog v8:
* removed mention of BATADV_UNICAST_TVLV from gen_batadv_offsets_v15(): currently a batman-adv unicast tvlv packets do not carry/support any payload frames, it only contains TVLVs, therefore removing it from offset considerations
* added a new patch to take variable length offsets, the tvlv length, in a batman-adv mcast packet into account ("batman-adv: Add offset support to batman-adv mcast packet"), allows using offset adjustments with this packet type, too

T-X commented:
Changelog v9:
* removed an implicit cast from size_t to bpf_u_int32 for the field_offset variable in gen_batadv_offsets_v15(), which was added in v8 in the newly added patch

simonwunderlich commented:
I would love to see this in tcpdump, too!

@mcr @fenner @infrastation @guyharris any chance you could help getting it merged? It sounded like @mcr was ready to merge it in 2021 already, but it's been stalled since then. Some embedded distributions already took the code as patch in to support their users, but having actually it merged would be great ...

infrastation commented:
Thank you for waiting. Rebased on the current master branch. I have looked through the changes one more time, there are minor places for improvement, but the most demanding parts are in `gencode.c` and `grammar.y.in`, as expected. One way to make it easier to verify could be capturing the new keywords in filter tests in `TESTrun`, I may be able to do that soon.

T-X commented:
@infrastation thanks once more for the feedback and a happy new year!
Btw. one more limitation I noticed, which I think other protocols in libpcap that use header offsets, like Geneve or PPPoE, would have too: I can't match multiple batman-adv packet types. Something like these won't work:
* `batadv 15 and not batadv 15 bcast`
* `batadv 15 bcast or batadv 15 iv-ogm`
* `(batadv 15 bcast) or (batadv 15 iv-ogm)`
* `batadv 15 (bcast or iv-ogm)` -> syntax error

But it seemed to me that this is a limitation of how these offsets work in libpcap right now in general, right? Once the keyword matches then the offset is applied immediately and globally and not within a specific scope? Which in the first three examples above would result in matching batman-adv packets within batman-adv packets, even when using brackets?

mcr commented:
T-X ***@***.***> wrote:
    > But it seemed to me that this is a limitation of how these offsets work
    > in libpcap right now in general, right? Once the keyword matches then
    > the offset is applied immediately and globally and not within a
    > specific scope? Which in the first three examples above would result in
    > matching batman-adv packets within batman-adv packets, even when using
    > brackets?

Correct.
And vlan is even more complex/annoying, because it can get taken out of the
packet into meta-data.

Fixing all of this requires some deep BPF knowledge, and some significant
reorganization to the pcap filter language.   And at least 3 weeks of
bikeshedding what are probably gonna have to be new keywords.

infrastation commented:
Three weeks of bikeshedding is not as bad as setting in stone something that causes pain by design, e.g. the `rnr` keyword.
